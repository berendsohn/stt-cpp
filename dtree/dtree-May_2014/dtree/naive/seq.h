// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
