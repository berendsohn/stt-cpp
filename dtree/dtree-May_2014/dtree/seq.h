// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
