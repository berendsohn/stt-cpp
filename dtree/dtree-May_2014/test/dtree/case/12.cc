#define DTREE_CONFIG EndSeq<WithReverse<WithReverse<WithAggr<dtree::Sum<int>, dtree::Index_<int>, WithAggr<dtree::Max<int>, dtree::GreaterEqual_<int>, WithValue<dtree::Nop<int>, WithAggr<dtree::Min<int>, dtree::Less_<int>, WithValue<dtree::Nop<int>, Begin > > > > > > > >
#include "test/dtree/seq.cc"
