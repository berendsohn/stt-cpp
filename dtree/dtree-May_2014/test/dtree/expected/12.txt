<<<
  5: v6 v2
  8: v7 v6
  15: v2 v1

  14: v3 v0
  1: v4 v6
  7: v6 v0
  10: v6 v5
  12: v2 v5
  6: v6 v1
  9: v5 v7
  11: v3 v3
  4: v5 v0
  13: v6 v7
  2: v7 v2
  3: v4 v0
>>>
<<<
    /5: v6 a6 v2 a2 a2 f1 f1
  @8: v7 a2 v6 a6 a7 f1 f1
    \15: v2 a2 v1 a1 a1 f0 f1

            /4: v5 a5 v0 a0 a0 f1 f0
          /13: v6 a5 v7 a7 a7 f1 f1
        /2: v7 a5 v2 a7 a9 f0 f1
      /11: v3 a3 v3 a7 a20 f1 f0
          /9: v5 a5 v7 a7 a7 f1 f0
        \6: v6 a5 v1 a7 a8 f0 f0
    /3: v4 a3 v0 a7 a20 f0 f1
  @12: v2 a2 v5 a6 a16 f0 f1
      /10: v6 a6 v5 a5 a5 f1 f0
    \7: v6 a3 v0 a6 a11 f0 f0
      \14: v3 a3 v0 a6 a6 f1 f0
        \1: v4 a4 v6 a6 a6 f0 f1
>>>
